### PR TITLE
chore: ga4 proxy test refactor

### PR DESCRIPTION
**Pull request by mihir-4116**
## What are the changes introduced in this PR?

Write a brief explainer on your code changes.

## What is the related Linear task?

Resolves INT-XXX

## Please explain the objectives of your changes below

Put down any required details on the broader aspect of your changes. If there are any dependent changes, **mandatorily** mention them here

### Any changes to existing capabilities/behaviour, mention the reason & what are the changes ?

N/A

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

N/A

### Any technical or performance related pointers to consider with the change?

N/A

@coderabbitai review

<hr>

### Developer checklist

- [ ] My code follows the style guidelines of this project

- [ ] **No breaking changes are being introduced.**

- [ ] All related docs linked with the PR?

- [ ] All changes manually tested?

- [ ] Any documentation changes needed with this change?

- [ ] Is the PR limited to 10 file changes?

- [ ] Is the PR limited to one linear task?

- [ ] Are relevant unit and component test-cases added?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/3137/test-report.html

File: test/integrations/destinations/ga4/dataDelivery/data.ts [hunk 1]
```diff
@@ -1,177 +1,3 @@
-export const data = [
```
Comment (utsabc) [inline test/integrations/destinations/ga4/dataDelivery/data.ts:1]: Lets not delete test cases for `v0`  along with these lets add `testScenariosV1API`

File: test/integrations/destinations/ga4/dataDelivery/business.ts [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+                metadata: {
+                  attemptNum: 1,
+                  destinationId: 'dummyDestinationId',
+                  dontBatch: false,
+                  jobId: 1,
+                  secret: {},
+                  sourceId: 'dummySourceId',
+                  userId: 'dummyUserId',
+                  workspaceId: 'dummyWorkspaceId',
+                },
```
Comment (utsabc) [inline test/integrations/destinations/ga4/dataDelivery/business.ts:138]: Use `generateMetdata` function from `testUtils.ts`

Comment (utsabc) [reply test/integrations/destinations/ga4/dataDelivery/business.ts:138]: Or use the same `metadata` variable declared above

File: test/integrations/destinations/ga4/dataDelivery/business.ts [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+        status: 200,
+        body: {
+          output: {
+            message:
+              'Validation Server Response Handler:: Validation Error for ga4 of field path :events | NAME_INVALID-Event at index: [0] has invalid name [campaign@details]. Only alphanumeric characters and underscores are allowed.',
+            response: [
+              {
+                error:
+                  'Validation Server Response Handler:: Validation Error for ga4 of field path :events | NAME_INVALID-Event at index: [0] has invalid name [campaign@details]. Only alphanumeric characters and underscores are allowed.',
+                metadata: {
```
Comment (utsabc) [inline test/integrations/destinations/ga4/dataDelivery/business.ts:183]: same here

File: test/integrations/destinations/ga4/dataDelivery/business.ts [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+  sourceId: 'dummySourceId',
+  destinationId: 'dummyDestinationId',
+  workspaceId: 'dummyWorkspaceId',
+  secret: {},
+  dontBatch: false,
+};
+
+const metadataArray = [proxyMetdata];
+
+export const testScenariosForV1API = [
```
Comment (utsabc) [inline test/integrations/destinations/ga4/dataDelivery/business.ts:95]: Use `ProxyV1TestData[]` for test cases type

File: test/integrations/destinations/ga4/dataDelivery/business.ts [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+                  jobId: 1,
+                  secret: {},
+                  sourceId: 'dummySourceId',
+                  userId: 'dummyUserId',
+                  workspaceId: 'dummyWorkspaceId',
+                },
+                statusCode: 400,
+              },
+            ],
+            statTags: {
```
Comment (utsabc) [inline test/integrations/destinations/ga4/dataDelivery/business.ts:196]: Check if the `statTags` can be declared as variables and reused

File: test/integrations/destinations/ga4/dataDelivery/business.ts [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+        status: 200,
+        body: {
+          output: {
+            message:
+              'Validation Server Response Handler:: Validation Error for ga4 of field path :undefined | INTERNAL_ERROR-Validation of item.price should prevent conversion from unsupported value [string_value: "$19"]',
+            response: [
+              {
+                error:
+                  'Validation Server Response Handler:: Validation Error for ga4 of field path :undefined | INTERNAL_ERROR-Validation of item.price should prevent conversion from unsupported value [string_value: "$19"]',
+                metadata: {
```
Comment (utsabc) [inline test/integrations/destinations/ga4/dataDelivery/business.ts:247]: Same for metdata

File: test/integrations/destinations/ga4/network.ts [hunk 7]
```diff
@@ -1,119 +1,139 @@
...
+const headers = {
+  HOST: 'www.google-analytics.com',
+  'Content-Type': 'application/json',
+};
+
+const params = {
+  api_secret: 'dymmyApiSecret',
+};
+
+const businessMockData = [
```
Comment (utsabc) [inline test/integrations/destinations/ga4/network.ts:10]: Keep existing mocks and label them join the new mocks with them

**Review by utsabc — CHANGES_REQUESTED** [review]

**Comment by utsabc** [comment]
https://github.com/rudderlabs/rudder-transformer/blob/457a18b2aec03aa0dfafcadc611b5f7176e97beb/test/integrations/destinations/braze/dataDelivery/other.ts

These scenarios also needs to be added

File: test/integrations/destinations/ga4/dataDelivery/data.ts (hunk 1 above)
Comment (mihir-4116) [reply test/integrations/destinations/ga4/dataDelivery/data.ts:1]: addressed

File: test/integrations/destinations/ga4/dataDelivery/business.ts (hunk 2 above)
Comment (mihir-4116) [reply test/integrations/destinations/ga4/dataDelivery/business.ts:138]: addressed

File: test/integrations/destinations/ga4/dataDelivery/business.ts (hunk 3 above)
Comment (mihir-4116) [reply test/integrations/destinations/ga4/dataDelivery/business.ts:183]: addressed

File: test/integrations/destinations/ga4/dataDelivery/business.ts (hunk 4 above)
Comment (mihir-4116) [reply test/integrations/destinations/ga4/dataDelivery/business.ts:95]: addressed

File: test/integrations/destinations/ga4/dataDelivery/business.ts (hunk 5 above)
Comment (mihir-4116) [reply test/integrations/destinations/ga4/dataDelivery/business.ts:196]: addressed

File: test/integrations/destinations/ga4/dataDelivery/business.ts (hunk 6 above)
Comment (mihir-4116) [reply test/integrations/destinations/ga4/dataDelivery/business.ts:247]: addressed

File: test/integrations/destinations/ga4/network.ts (hunk 7 above)
Comment (mihir-4116) [reply test/integrations/destinations/ga4/network.ts:10]: addressed

**Comment by mihir-4116** [comment]
> https://github.com/rudderlabs/rudder-transformer/blob/457a18b2aec03aa0dfafcadc611b5f7176e97beb/test/integrations/destinations/braze/dataDelivery/other.ts
> 
> These scenarios also needs to be added

not needed, we will never get these responses from ga4. we will always receive 200 only

File: test/integrations/destinations/ga4/dataDelivery/business.ts [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+      response: {
+        status: 400,
+        body: {
+          output: {
+            destinationResponse:
+              'Validation of item.price should prevent conversion from unsupported value [string_value: "$19"]',
+            message:
+              'Validation Server Response Handler:: Validation Error for ga4 of field path :undefined | INTERNAL_ERROR-Validation of item.price should prevent conversion from unsupported value [string_value: "$19"]',
+            statTags: {
+              destType: 'GA4',
```
Comment (koladilip) [inline test/integrations/destinations/ga4/dataDelivery/business.ts:207]: can we reuse "expected stat tags"

Comment (mihir-4116) [reply test/integrations/destinations/ga4/dataDelivery/business.ts]: addressed

File: test/integrations/destinations/ga4/dataDelivery/business.ts [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+      response: {
+        status: 200,
+        body: {
+          output: {
+            message:
+              'Validation Server Response Handler:: Validation Error for ga4 of field path :undefined | INTERNAL_ERROR-Validation of item.price should prevent conversion from unsupported value [string_value: "$19"]',
+            response: [
+              {
+                error:
+                  'Validation Server Response Handler:: Validation Error for ga4 of field path :undefined | INTERNAL_ERROR-Validation of item.price should prevent conversion from unsupported value [string_value: "$19"]',
```
Comment (koladilip) [inline test/integrations/destinations/ga4/dataDelivery/business.ts:339]: repeated

Comment (mihir-4116) [reply test/integrations/destinations/ga4/dataDelivery/business.ts:339]: no, it's not repeated. error type is different. one is for invalid event name and other is for invalid parameter

Comment (koladilip) [reply test/integrations/destinations/ga4/dataDelivery/business.ts:339]: image

File: test/integrations/destinations/ga4/dataDelivery/business.ts [hunk 10]
```diff
@@ -244,29 +337,11 @@ export const testScenariosForV1API = [
...
-                  attemptNum: 1,
-                  destinationId: 'dummyDestinationId',
-                  dontBatch: false,
-                  jobId: 1,
-                  secret: {},
-                  sourceId: 'dummySourceId',
-                  userId: 'dummyUserId',
-                  workspaceId: 'dummyWorkspaceId',
-                },
+                metadata: generateMetadata(1),
```
Comment (utsabc) [inline test/integrations/destinations/ga4/dataDelivery/business.ts:340]: Incorrect succesCriteria

**Review by utsabc — CHANGES_REQUESTED** [review]

File: test/integrations/destinations/ga4/dataDelivery/business.ts (hunk 10 above)
Comment (mihir-4116) [reply test/integrations/destinations/ga4/dataDelivery/business.ts:340]: addressed

File: test/integrations/destinations/ga4/dataDelivery/business.ts [hunk 11]
```diff
@@ -180,29 +291,11 @@ export const testScenariosForV1API = [
...
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply test/integrations/destinations/ga4/dataDelivery/business.ts:294]: addressed

File: test/integrations/destinations/ga4/dataDelivery/business.ts [hunk 12]
```diff
@@ -79,20 +83,136 @@ const invalidParameterValueRequest = {
...
+        },
+      },
+    },
+  },
+  {
+    id: 'ga4_v0_scenario_3',
+    name: 'ga4',
+    description:
+      '[Proxy v0 API] :: Test for a invalid parameter value - where the destination responds with 200 with error for invalid parameter value',
+    successCriteria: 'Should return 200 with error with destination response',
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply test/integrations/destinations/ga4/dataDelivery/business.ts]: addressed

File: test/integrations/destinations/ga4/dataDelivery/business.ts [hunk 13]
```diff
@@ -79,20 +83,136 @@ const invalidParameterValueRequest = {
...
+        },
+      },
+    },
+  },
+  {
+    id: 'ga4_v0_scenario_2',
+    name: 'ga4',
+    description:
+      '[Proxy v0 API] :: Test for a invalid event name - where the destination responds with 200 with error for invalid event name',
+    successCriteria: 'Should return 200 with error with destination response',
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply test/integrations/destinations/ga4/dataDelivery/business.ts]: addressed

File: test/integrations/destinations/ga4/dataDelivery/business.ts (hunk 9 above)
Comment (koladilip) [reply test/integrations/destinations/ga4/dataDelivery/business.ts:339]: is it possible keep them as variables

Comment (mihir-4116) [reply test/integrations/destinations/ga4/dataDelivery/business.ts]: addressed

**Review by utsabc — APPROVED** [review]

File: test/integrations/destinations/ga4/dataDelivery/business.ts [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+        method: 'POST',
+      },
+    },
+    output: {
+      response: {
+        status: 400,
+        body: {
+          output: {
+            destinationResponse:
+              'Validation of item.price should prevent conversion from unsupported value [string_value: "$19"]',
```
Comment (koladilip) [inline test/integrations/destinations/ga4/dataDelivery/business.ts:207]: create a constant for this

Comment (mihir-4116) [reply test/integrations/destinations/ga4/dataDelivery/business.ts]: addressed

**Comment by mihir-4116** [comment]
> Please add the count of new scenarios added if not put NA

NA

**Review by utsabc — APPROVED** [review]

**Pull request merged**